File: src/com/suushiemaniac/cubing/alglib/move/MegaminxMove.java
package com.suushiemaniac.cubing.alglib.move;

import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxUpModifier;
import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxWideModifier;
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxPlane;
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxUpPlane;
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxWidePlane;
import com.suushiemaniac.cubing.alglib.transform.Transform;

public class MegaminxMove implements Move {
    private final MegaminxPlane plane;
    private final MegaminxModifier modifier;

    public MegaminxMove(MegaminxWidePlane plane, MegaminxWideModifier modifier) {
        this.plane = plane;
        this.modifier = modifier;
    }

    public MegaminxMove(MegaminxUpPlane plane, MegaminxUpModifier modifier) {
        this.plane = plane;
        this.modifier = modifier;
    }

    @Override
    public boolean merges(Move other) {
        return false;
    }

    @Override
    public boolean cancels(Move other) {
        return other instanceof MegaminxMove
                && this.plane == other.getPlane()
                && this.modifier == other.getModifier().inverse();
    }

    @Override
    public boolean mayAppend(Move other) {
        return true;
    }

    @Override
    public MegaminxPlane getPlane() {
        return this.plane;
    }

    @Override
    public MegaminxModifier getModifier() {
        return this.modifier;
    }

    @Override
    public MegaminxMove merge(Move other) {
        return null;
    }

    @Override
    public int getDepth() {
        return this.plane instanceof MegaminxUpPlane ? 1 : 6;
    }

    @Override
    public MegaminxMove inverse() {
        if (!this.plane.getType().equals(this.modifier.getType())) return null;
        if (this.plane instanceof MegaminxWidePlane && this.modifier instanceof MegaminxWideModifier)
            return new MegaminxMove((MegaminxWidePlane) this.plane, (MegaminxWideModifier) this.modifier.inverse());
        else if (this.plane instanceof MegaminxUpPlane && this.modifier instanceof MegaminxUpModifier)
            return new MegaminxMove((MegaminxUpPlane) this.plane, (MegaminxUpModifier) this.modifier.inverse());
        else return null;
    }

    @Override
    public Move transform(Transform transform) {
        return transform.supportsMoveClass(this) ? transform.transform(this) : this;
    }

    @Override
    public String toFormatString() {
        return this.plane.toFormatString() + this.modifier.toFormatString();
    }
}

File: src/main/kotlin/com/suushiemaniac/cubing/alglib/move/MegaminxMove.kt
package com.suushiemaniac.cubing.alglib.move

import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxModifier
import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxUpModifier
import com.suushiemaniac.cubing.alglib.move.modifier.MegaminxWideModifier
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxPlane
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxUpPlane
import com.suushiemaniac.cubing.alglib.move.plane.MegaminxWidePlane

data class MegaminxMove(override val plane: MegaminxPlane, override val modifier: MegaminxModifier) : Move {
    override val notation = "${this.plane.notation}${this.modifier.notation}"
    override val depth = if (this.plane is MegaminxUpPlane) 1 else 6

    override fun cancels(other: Move): Boolean {
        return (other is MegaminxMove
                && this.plane === other.plane
                && this.modifier === other.modifier.inverse())
    }

    override fun mayAppend(other: Move) = other is MegaminxMove
    override fun merges(other: Move) = false
    override fun merge(other: Move) = other as MegaminxMove

    override fun inverse(): MegaminxMove {
        if (this.plane.type != this.modifier.type) return this

        return if (this.plane is MegaminxWidePlane && this.modifier is MegaminxWideModifier)
            MegaminxMove(this.plane, this.modifier.inverse())
        else if (this.plane is MegaminxUpPlane && this.modifier is MegaminxUpModifier)
            MegaminxMove(this.plane, this.modifier.inverse())
        else
            this
    }

    override fun toString() = this.notation
}
